storage/store/headers.go

```diff
@@ -21,34 +21,83 @@ type Headers struct {
+	if chainID.IsClusterChain() {
+		panic("NewHeaders called on cluster chain ID - use NewClusterHeaders instead")
 	}
-
 	retrieveHeight := func(r storage.Reader, height uint64) (flow.Identifier, error) {
 		var id flow.Identifier
 		err := operation.LookupBlockHeight(r, height, &id)
 		return id, err
 	}
-
 	retrieveView := func(r storage.Reader, view uint64) (flow.Identifier, error) {
 		var id flow.Identifier
 		err := operation.LookupCertifiedBlockByView(r, view, &id)
 		return id, err
 	}
+	return newHeaders(collector, db, chainID, retrieveHeight, retrieveView)
+}
+
+// NewClusterHeaders creates a Headers instance for a collection cluster chain, which stores block headers for cluster blocks.
+// It supports storing, caching and retrieving by block ID, and additionally an index by header height.
+func NewClusterHeaders(collector module.CacheMetrics, db storage.DB, chainID flow.ChainID) *Headers {
+	if !chainID.IsClusterChain() {
+		panic("NewClusterHeaders called on non-cluster chain ID - use NewHeaders instead")
+	}
+	retrieveHeight := func(r storage.Reader, height uint64) (flow.Identifier, error) {
+		var id flow.Identifier
+		err := operation.LookupClusterBlockHeight(r, chainID, height, &id)
+		return id, err
+	}
+	retrieveView := func(r storage.Reader, height uint64) (flow.Identifier, error) {
+		var id flow.Identifier
+		return id, fmt.Errorf("retrieve by view not implemented for cluster headers")
+	}
+	return newHeaders(collector, db, chainID, retrieveHeight, retrieveView)
+}
+
+// newHeaders contains shared logic for Header storage, including storing and retrieving by block ID
+func newHeaders(collector module.CacheMetrics,
+	db storage.DB,
+	chainID flow.ChainID,
+	retrieveHeight retrieveFunc[uint64, flow.Identifier],
+	retrieveView retrieveFunc[uint64, flow.Identifier],
+) *Headers {
+	storeWithLock := func(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, header *flow.Header) error {
+		if header.ChainID != chainID {
+			return fmt.Errorf("expected chain ID %v, got %v: %w", chainID, header.ChainID, storage.ErrWrongChain)
+		}
+		if chainID.IsClusterChain() {
+			if !lctx.HoldsLock(storage.LockInsertOrFinalizeClusterBlock) {
+				return fmt.Errorf("missing lock: %v", storage.LockInsertOrFinalizeClusterBlock)
+			}
+		} else {
+			if !lctx.HoldsLock(storage.LockInsertBlock) {
+				return fmt.Errorf("missing lock: %v", storage.LockInsertBlock)
+			}
+		}
+		return operation.InsertHeader(lctx, rw, blockID, header)
+	}
+
+	retrieve := func(r storage.Reader, blockID flow.Identifier) (*flow.Header, error) {
+		var header flow.Header
+		err := operation.RetrieveHeader(r, blockID, &header)
+		if err != nil {
+			return nil, err
+		}
+		// raise an error when the retrieved header is for a different chain than expected
+		if header.ChainID != chainID {
+			return nil, fmt.Errorf("expected chain ID %v, got %v: %w", chainID, header.ChainID, storage.ErrWrongChain)
+		}
```

Comment:
Generic errors indicate critical errors, from the perspective of the _called code_ ([our Coding Conventions](https://github.com/onflow/flow-go/blob/master/CodingConventions.md) discusses this in more detail). This is by no means a critical error from the perspective of `Headers`. Hence it should return a sensible sentinel, the sentinel should be documented as part of the API and covered by test cases.